drivers/led_strip/ws2812_uart.c

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2024 Croxel, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT worldsemi_ws2812_uart
+
+#include <zephyr/drivers/led_strip.h>
+
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_uart);
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/math_extras.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/dt-bindings/led/led.h>
+
+struct ws2812_uart_cfg {
+	const struct device *uart_dev; /*UART device*/
+	uint8_t *px_buf;
+	size_t px_buf_size;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	uint16_t reset_delay;
+};
+
+/* This lookup table maps 3 bits of information into (7+startBit+stopBit = 9)-bits,*/
+/* which will be used to send ws2812 rgb led */
+static const uint8_t ws2812_uart_lut[8] = {
+	0b1011011,
+	0b0011011,
+	0b1010011,
+	0b0010011,
+	0b1011010,
+	0b0011010,
+	0b1010010,
+	0b0010010,
+};
+
+/*
+ * Serialize an 24-bit color channel value into an equivalent sequence
+ * of 8 bytes of uart frames,.
+ */
+static inline void ws2812_uart_ser(uint8_t *buf, uint8_t color[3])
+{
+	int i, j;
+	int buff_indx = 0;
+	uint8_t serialized_val = 0;
+
+	for (i = 0; i < 3; i++) {
+		for (j = 0; j < 8; j++) {
+			serialized_val = (serialized_val << 1) | ((color[i]>>(7-j))&0x01);
+			if (((i*8)+j) % 3 == 2) {
+				buf[buff_indx++] = ws2812_uart_lut[serialized_val];
+				serialized_val = 0;
+			}
+		}
+	}
+}
+
+/*
+ * Returns true if and only if cfg->px_buf is big enough to convert
+ * num_pixels RGB color values into UART frames.
+ */
+static inline bool num_pixels_ok(const struct ws2812_uart_cfg *cfg,
+				 size_t num_pixels)
+{
+	size_t nbytes;
+	bool overflow;
+
+	overflow = size_mul_overflow(num_pixels, cfg->num_colors * 8, &nbytes);
+	return !overflow && (nbytes <= cfg->px_buf_size);
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
```

Comment:
seems like this function can go and just use k_usleep directly?

Reply:
Yeah thats what I was thinking too because I used the ws2812 spi driver as a reference and this function was there.
Removing it now.